tokio/src/time/driver/mod.rs

```diff
@@ -145,189 +182,255 @@ where
+
+        Ok(())
     }
+}
+
+impl Handle {
+    /// Runs timer related logic, and returns the next wakeup time
+    pub(self) fn process(&self) -> Option<NonZeroU64> {
+        let now = self.time_source().now();
 
-    fn clear_entry(&mut self, entry: &Arc<Entry>) {
-        self.wheel.remove(entry);
-        entry.set_when_internal(None);
+        self.process_at_time(now)
     }
 
-    /// Fires the entry if it needs to, otherwise queue it to be processed later.
-    fn add_entry(&mut self, entry: Arc<Entry>, when: u64) {
-        use crate::time::error::InsertError;
+    pub(self) fn process_at_time(&self, now: u64) -> Option<NonZeroU64> {
+        let mut any_awoken = false;
 
-        entry.set_when_internal(Some(when));
+        let mut waker_list: [Option<Waker>; 32] = Default::default();
+        let mut waker_idx = 0;
 
-        match self.wheel.insert(when, entry) {
-            Ok(_) => {}
-            Err((entry, InsertError::Elapsed)) => {
-                // The entry's deadline has elapsed, so fire it and update the
-                // internal state accordingly.
-                entry.set_when_internal(None);
-                entry.fire(when);
-            }
-            Err((entry, InsertError::Invalid)) => {
-                // The entry's deadline is invalid, so error it and update the
-                // internal state accordingly.
-                entry.set_when_internal(None);
-                entry.error(Error::invalid());
-            }
-        }
-    }
-}
+        let mut lock = self.lock();
 
-impl<T> Park for Driver<T>
-where
-    T: Park,
-{
-    type Unpark = T::Unpark;
-    type Error = T::Error;
+        assert!(now >= lock.elapsed);
 
-    fn unpark(&self) -> Self::Unpark {
-        self.park.unpark()
-    }
+        while let Some(entry) = lock.wheel.poll(now) {
+            debug_assert!(unsafe { entry.is_pending() });
 
-    fn park(&mut self) -> Result<(), Self::Error> {
-        self.process_queue();
+            // SAFETY: We hold the driver lock, and just removed the entry from any linked lists.
+            if let Some(waker) = unsafe { entry.fire(Ok(())) } {
+                any_awoken = true;
 
-        match self.wheel.poll_at() {
-            Some(when) => {
-                let now = self.clock.now();
-                let deadline = self.expiration_instant(when);
+                waker_list[waker_idx] = Some(waker);
 
-                if deadline > now {
-                    let dur = deadline - now;
+                waker_idx += 1;
 
-                    if self.clock.is_paused() {
-                        self.park.park_timeout(Duration::from_secs(0))?;
-                        self.clock.advance(dur);
-                    } else {
-                        self.park.park_timeout(dur)?;
+                if waker_idx == waker_list.len() {
+                    // Wake a batch of wakers. To avoid deadlock, we must do this with the lock temporarily dropped.
+                    std::mem::drop(lock);
```

Comment:
nit: `drop` is in the prelude:
```suggestion
                    drop(lock);
```

Reply:
huh. TIL.